shared/src/metabase/shared/models/visualization_settings.cljc

```diff
@@ -0,0 +1,410 @@
+  {:added "0.40.0"}
+  [field-id]
+  {::field-id field-id})
+
+(s/def ::field-id integer?)
+
+(defn column-ref-for-column-name
+  "Creates a normalized column ref map for the given `col-name`. This becomes a key in the `::column-settings` map."
+  {:added "0.40.0"}
+  [col-name]
+  {::column-name col-name})
+
+(s/def ::column-name string?)
+
+(defn column-ref-for-qualified-name
+  "Creates a normalized column ref map for the given fully qualified name (string). This becomes the key in the
+  `::column-settings` map."
+  {:added "0.40.0"}
+  [field-qualified-name]
+  {::field-qualified-name field-qualified-name})
+
+(s/def ::field-id string?)
+
+(defn keyname
+  "Returns the full string name of the keyword `kw`, including any \"namespace\" portion followed by forward slash.
+
+  From https://clojuredocs.org/clojure.core/name#example-58264f85e4b0782b632278bf
+  Clojure interprets slashes as keyword/name separators, so we need to do something hacky to get the \"full\" name here
+  because our \"keyword value\" (as parsed from JSON/YAML/etc.) might actually look like the string version of a
+  Clojure vector, which itself can contain a fully qualified name for serialization"
+  {:added "0.40.0"}
+  [kw]
+  (str (if-let [kw-ns (namespace kw)] (str kw-ns "/")) (name kw)))
+
+(defn- parse-json-string
+  "Parse the given `json-str` to a map. In Clojure, this uses Cheshire. In Clojurescript, it uses `cljs.reader`."
+  [json-str]
+  #?(:clj  (json/parse-string json-str)
+     :cljs (cljs-reader/read-string json-str)))
+
+(defn- encode-json-string
+  "Encode the given `obj` map as a JSON string. In Clojure, this uses Cheshire. In Clojurescript, it uses
+  `cljs.core.clj->js` in conjunction with `cljs.js`."
+  [obj]
+  #?(:clj  (json/encode obj)
+     :cljs (.stringify js/JSON (clj->js obj))))
+
+(defn parse-column-ref
+  "Opposite of the `column-ref-for-*` fns. Converts the given string representation of a column settings key into its
+  normalized form. The `col-ref` argument can be a string or keyword (which is produced by YAML parsing, for instance).
+  In case it is a keyword, it will be converted to its full name. \"Full\" means that the portions before and after any
+  slashes will be included verbatim (via the `keyname` helper fn). This is necessary because our serialization code
+  considers a forward slash to be a legitimate portion of a fully qualified name, whereas Clojure considers it to be a
+  namespace/local name separator."
+  {:added "0.40.0"}
+  [col-ref]
+  (let [parts           ((comp vec parse-json-string) (if (keyword? col-ref) (keyname col-ref) col-ref))
+        [kind & params] parts]
+    (case kind
+      "ref"  (let [[[t id & _]] params]
+               (case t "field" (cond (int? id)    {::field-id id}
+                                     (string? id) {::field-qualified-name id})))
+      "name" {::column-name (first params)})))
```

Comment:
Why not just use Spec to do all of this. Seems like it would be a lot easier to write a spec for the shape and have it tell you what it got

Reply:
I have some specs defined but I'm unclear on what you mean here.  Do you mean using `conform`?

Reply:
@camsaul , can you clarify on this one?  Doing a big cleanup/PR feedback commit and I wanted to understand this better.

Reply:
OK, have a look at the changes now to see if that's what you had in mind.